### PR TITLE
Introduced Data Transformer

### PR DESCRIPTION
**Pull request by integraledelebesgue**
**Stack**:
- #2555
- #2548
- #2547
- #2546
- #2545

### CODE REVIEW

File: docs/src/starknet/calldata-transformation.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  * `ContractAddress`
+  * `ClassHash`
+  * `StorageAddress`
+  * `EthAddress`
+  * `bytes31`
+  * `Array` - using `array![]` macro
+
+Numeric types (primitives and `felt252`) can be paseed with type suffix specified - for example `--calldata 420_u64`.
+
+It's **not allowed** to write any function call expressions, either builtin, or user-defined.
```
Comment (kkawula) [inline docs/src/starknet/calldata-transformation.md:107]: I would make it more verbose, the message of this line isn't obvious to me

File: docs/src/starknet/calldata-transformation.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+### More complex examples
+
+  1. `complex_fn` - different data types:
+
+  ```shell
+  $ sncast --account myuser \
+      call \
+      --url http://127.0.0.1:5050 \
+      --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+      --function tuple_fn \
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:118]: I think it should be `complex_fn`

File: docs/src/starknet/calldata-transformation.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+$ sncast --account myuser \
+  call \
+  --url http://127.0.0.1:5050 \
+  --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+  --function u256_fn \
+  --calldata 0x10 \
+  --block-id latest
+```
+
+Despite the call data **not being** the proper serialization of an `u256` (which would be `0x10 0x0`), call succeeds.
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:171]: Maybe this should be a bit more strongly worded e.g. explicitly warn users to always use type suffixes etc. so their data doesn't become serialized incorrectly

Comment (cptartur) [reply docs/src/starknet/calldata-transformation.md:171]: Also it would probably make sense to add a flag that disables data transformation just to be extra sure. This could be a separate issue and tackled by someone else though

File: docs/src/starknet/index.md [hunk 4]
```diff
@@ -43,6 +43,22 @@ response: [0x0]
...
+### Calldata
+
+Some `sncast` commands (namely `call`, `deploy` and `invoke`) allow passing *calldata* - a series of arguments to perform an action with on blockchain.
+
+In the example above we called a function with an argument: `0x0`, passed using `--calldata` flag.
+
+Please note the notation of the argument. The default way of passing calldata is a list of hexadecimally encoded field elements - the *serialized* calldata.
+To obtain the serialized form of the wished data, one must write a Cairo program calling `Serde::serialize` on subsequent arguments and displaying the results.
+
+It is also possible to pass calldata in more friendly, human readable form due to the [calldata transformation](./calldata-transformation.md) feature present in Cast.
```
Comment (cptartur) [inline docs/src/starknet/index.md:56]:
```suggestion
It is also possible to pass calldata in more friendly, human readable form thanks to the [calldata transformation](./calldata-transformation.md) feature present in Cast.
```

File: docs/src/starknet/index.md [hunk 5]
```diff
@@ -43,6 +43,22 @@ response: [0x0]
...
+
+> ⚠️ **Warning**
+> Cast will not werify the serialized calldata. Any errors caused by passing improper calldata in a serialized form will originate from the network.
```
Comment (cptartur) [inline docs/src/starknet/index.md:59]:
```suggestion
> Cast will not verify the serialized calldata. Any errors caused by passing improper calldata in a serialized form will originate from the network.
```

File: docs/src/starknet/calldata-transformation.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  2. `complex_struct_fn` - struct nesting:
+
+  ```shell
+  $ sncast --account myuser \
+      call \
+      --url http://127.0.0.1:5050 \
+      --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+      --function tuple_fn \
+      --calldata \
+          data_transformer_contract::ComplexStruct { \
```
[the remark this answers is not in the document]
Comment (integraledelebesgue) [reply docs/src/starknet/calldata-transformation.md:139]: It should be `NestedStructWithField` instead, my fault

File: docs/src/starknet/calldata-transformation.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+#[starknet::contract]
+pub mod DataTransformerContract {
+    /* ... */
+
+    use super::*;
+
+    fn tuple_fn(self: @ContractState, a: (felt252, u8, Enum)) { ... }
+
+    fn complex_struct_fn(self: @ContractState, a: ComplexStruct) { ... }
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:35]:
```suggestion
    fn nested_struct_fn(self: @ContractState, a: NestedStructWithField) { ... }
```

File: docs/src/starknet/calldata-transformation.md [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+> 📝 **Note**
+> User-defined items such as enums and structs should be referred to depending on a way they are defined in ABI.
+> In general, paths to items have form: `<project-name>::<module-path>::<item-name>`.
+
+### Supported expressions
+
+> 💡 **Info**
+> Only **constant** expressions are supported. Defining and referencing variables and calling functions (either builtin, user-defined or external) is not allowed.
+
+Cast supports most important Cairo corelib types:
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:95]:
```suggestion
Cast supports converting most important Cairo corelib types:
```

File: docs/src/starknet/calldata-transformation.md [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+    --function tuple_fn \
+    --calldata 0x10 0x3 0x0 \
+    --block-id latest
+```
+
+However, `sncast` allows passing the data in far more handy, human-readable form - as Cairo expressions.
+When calldata is delivered in such form, Cast will perform serialization automatically, based on an ABI of the contract we interact with.
+
+### Basic example
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:67]:
```suggestion
## Basic Example
```

File: docs/src/starknet/calldata-transformation.md [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+getting the same result.
+
+> 📝 **Note**
+> All data types are serialized according to the official [Starknet specification](https://docs.starknet.io/architecture-and-concepts/smart-contracts/serialization-of-cairo-types/).
+
+> 📝 **Note**
+> User-defined items such as enums and structs should be referred to depending on a way they are defined in ABI.
+> In general, paths to items have form: `<project-name>::<module-path>::<item-name>`.
+
+### Supported expressions
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:90]:
```suggestion
## Supported Expressions
```

File: docs/src/starknet/calldata-transformation.md [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+  * `ContractAddress`
+  * `ClassHash`
+  * `StorageAddress`
+  * `EthAddress`
+  * `bytes31`
+  * `Array` - using `array![]` macro
+
+Numeric types (primitives and `felt252`) can be paseed with type suffix specified - for example `--calldata 420_u64`.
+
+### More complex examples
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:110]:
```suggestion
## More Complex Examples
```

File: docs/src/starknet/calldata-transformation.md [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+      --function nested_struct_fn \
+      --calldata \
+          data_transformer_contract::NestedStructWithField { \
+              a: data_transformer_contract::SimpleStruct { a: 10 }, \
+              b: 12 \
+          } \
+      --block-id latest
+  ```
+
+### Caveats
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:147]: Unless it was meant to be a subsection

```suggestion
## Caveats
```

File: docs/src/starknet/calldata-transformation.md [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+#[starknet::contract]
+pub mod DataTransformerContract {
+    /* ... */
+
+    use super::*;
+
+    fn tuple_fn(self: @ContractState, a: (felt252, u8, Enum)) { ... }
+
+    fn complex_struct_fn(self: @ContractState, a: ComplexStruct) { ... }
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:35]:
```suggestion
    fn nested_struct_fn(self: @ContractState, a: NestedStructWithField) { ... }
```

File: docs/src/starknet/calldata-transformation.md [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+    --calldata 0x10 0x3 0x0 \
+    --block-id latest
+```
+
+However, `sncast` allows passing the data in far more handy, human-readable form - as a series of Cairo expressions.
+When calldata is delivered in such form, Cast will perform serialization automatically, based on an ABI of the contract we interact with.
+
+To transform Cairo expressions to serialized form, calldata must meet the following requirements:
+
+* Arguments are comma-separated (on a contrary to a standard serialized input which is whitespace-separatd). Note the lack of commas in the example above - it signals Cast that the input is serialized.
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:69]: This could maybe be a note or a warning so it's more explicit

File: docs/src/starknet/calldata-transformation.md [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+* If only one argument without a comma is given, it should be a *nontrivial* literal (e.g. a struct or enum constructor).
+  It cannot be a simple numeric litral - `--calldata 0x2137` will be treated as serialized.\
+  To transform such, we need to add a type suffix - for example: `--calldata 0x2137_u96`. See [Caveats](./calldata-transformation.md#caveats)
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:73]: Would a trailing comma after expression work as well?

File: docs/src/starknet/calldata-transformation.md [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+      --block-id latest
+  ```
+
+## Caveats
+
+> 💡 **Info**
+> To force transformation of a single numeric literal, simply add a type suffix to it.
+
+> 💡 **Info**
+> To force transformation of a list of `felt252`, separate its elements with commas.
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:166]: I think these could be subsection in here, we have a bunch of info and similar in this page already

File: docs/src/starknet/calldata-transformation.md [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+  call \
+  --url http://127.0.0.1:5050 \
+  --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+  --function u256_fn \
+  --calldata 0x10 \
+  --block-id latest
+```
+
+Despite the call data being the proper literal of an `u256`, call fails.
+That's because Cast interpreted `0x10` as a hex-encoded `felt252`, not `u256`. A properly serialized `u256` should have two `felt252` components.
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:192]: I'd reword it and just say that cast will always treat the data as serialized unless explicitly told to use data transformer (with commas, suffixes etc). I appreciate the detailed explanation but average user will probably just skim through this page if anything.

File: docs/src/starknet/calldata-transformation.md
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:1]: My only caveat with this page is that it's quite in depth at times where some kind of maybe a simple explanation would also be nice to have.
I think we can improve that in a separate issue though.

**Review by cptartur — COMMENTED** [review]
See my unresolved comments from previous review as well

File: docs/src/starknet/calldata-transformation.md (hunk 17 above)
Comment (integraledelebesgue) [reply docs/src/starknet/calldata-transformation.md:192]: This section is meant to be in-depth and optional to read for a better understanding. I reworked some ohter sections to emphasize the transformation rules and make them more straightforward.

File: docs/src/starknet/calldata-transformation.md (hunk 15 above)
Comment (integraledelebesgue) [reply docs/src/starknet/calldata-transformation.md:73]: Yeah, trailing comma works. I modified one of the test cases to verify it.

File: docs/src/starknet/calldata-transformation.md [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+    --function tuple_fn \
+    --calldata 0x10 0x3 0x0 \
+    --block-id latest
+```
+
+However, `sncast` allows passing the data in far more handy, human-readable form - as a series of Cairo expressions.
+When calldata is delivered in such form, Cast will perform serialization automatically, based on an ABI of the contract we interact with.
+
+> 📝 **Note**
+> Calldata transformation becomes engaged when following requirements are met:
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:68]:
```suggestion
> Calldata transformation is enabled only when following requirements are met:
```

File: docs/src/starknet/calldata-transformation.md [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+>
+> * **Arguments are comma-separated** (on a contrary to a standard serialized input which is whitespace-separatd). Note the lack of commas in the example above - it signals Cast that the input is serialized.
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:70]:
```suggestion
> * **Arguments are comma-separated** (on a contrary to a standard serialized input which is whitespace-separated). Note the lack of commas in the example above - it signals Cast that the input is serialized.
```

File: docs/src/starknet/calldata-transformation.md [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+$ sncast --account myuser \
+    call \
+    --url http://127.0.0.1:5050 \
+    --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+    --function tuple_fn \
+    --calldata (0x10, 3, data_stransformer_contract::Enum::One) \
+    --block-id latest
+```
+
+getting the same result.
```
Comment (cptartur) [inline docs/src/starknet/calldata-transformation.md:93]:
```suggestion
getting the same result. Note that values are separated by commas `,`.
```

I know this is repeated but I think it might be worth mentioning again.

**Review by cptartur — APPROVED** [review]
